### ISSUE TITLE
Fetch from Ensembl is failing

### ISSUE THREAD

**Issue by mbookman** [googlegenomics/api-client-python#11]
main.py references the ensemble server as:

SUPPORTED_BACKENDS = {
  'Ensembl' : {'name': 'Ensembl',
            'url': 'http://193.62.52.232:8081/%s?%s',
            'datasets': {'1000 genomes pilot 2': '2'}}
}

However this server does not respond:

$ curl http://193.62.52.232:8081
curl: (7) Failed to connect to 193.62.52.232 port 8081: Connection timed out

**Comment by calbach** [comment]
@sarahhunt do you know the status of the Ensembl implementation of the v0.5.1 Variants API (or do you know who would know)? I'm not surprised that this hardcoded IP address broke, but is there an implementation still accessible somewhere? Thanks.

**Comment by sarahhunt** [comment]
@calbach - we no longer have v0.5.1 running - we have v0.6.0 on our production server http://rest.ensembl.org/ga4gh. 
We could look at putting one up next month, if it would be useful? 
The API isn't getting much traffic, so we are planning to use the main site to let people try the latest version. Supporting the last 'popular' version also makes sense.

**Comment by calbach** [comment]
Thanks Sarah. Lets try out 0.6.0 first and see if that works. IIIRC there were no breaking changes to the Variants API between these two versions.

I guess it is a potential concern if the version hosted @ http://rest.ensembl.org/ga4gh changes on us, but we can probably punt on that.

**Comment by pgrosu** [comment]
@calbach I think you might need the Readsets as well for GABrowse.  It might be good to check the documentation first at the link below (the GA4GH section seems to be called "Variation GA4GH"):

http://rest.ensembl.org/

By the way, the only way I found to extract the CallSet Names is by the following, but you'll have to go through the data:


 curl 'http://rest.ensembl.org/ga4gh/variants/search' -H 'Content-type:application/json' -H 'Accept:application/json' -X POST -d '{ "variantSetId": 1, "referenceName": 22,"start": 17190024 ,"end":  17671934 ,  "pageToken":"", "pageSize": 1 }'  | sed 's/,/\n/g' | grep callSetName | cut -f4 -d\" | sort | uniq

The only dataset I found to be populated was `1000 Genomes phase3`:

curl 'http://rest.ensembl.org/ga4gh/datasets/search' -H 'Content-type:application/json'     -H 'Ac' -X POST -d '{ "pageSize": 3 }'  | sed 's/,/\n\n/g'

```
{"datasets":[{"id":"6e340c4d1e333c7a676b1710d2e3953c"

"description":"1000 Genomes phase3"}]

"nextPageToken":null}
```

Hope it helps,
~p

**Comment by pgrosu** [comment]
When I tried the following for Reads based on the [search method definition](https://github.com/ga4gh/schemas/blob/master/src/main/resources/avro/readmethods.avdl#L8-56) it seems not to be implemented - maybe for Reads there might be another link?:

curl 'http://rest.ensembl.org/ga4gh/reads/search' -H 'type:application/json'  -H 'Accept:application/json' -X POST -d '{}'

{"error":"page not found. Please check your uri and refer to our documentation http://rest.ensembl.org/"}

**Comment by calbach** [comment]
I'm pretty sure Ensembl does not implement the Reads API and never has. Accessing the Reads API should not be a requirement for fixing this GitHub issue.

**Comment by mbookman** [comment]
I think that there is a bug in the callsets/search function where it requires a pageSize to be specified. I have not observed this as a requirement for other calls.

This works:

```
$ curl 'http://rest.ensembl.org/ga4gh/callsets/search?fields=callSets(id,name))' -H 'Content-type:application/json;chareset=UTF-8' -H 'Accept:application/json' -X POST -d '{ "variantSetId": "1", "name": "", "pageToken":"", "pageSize": 1 }'
{"nextPageToken":1,"callSets":[{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00096","name":"HG00096","updated":1432745640000,"id":"1:HG00096"}]}


```

This returns nothing:

but if I remove the pageSize:

```
$ curl 'http://rest.ensembl.org/ga4gh/callsets/search?fields=callSets(id,name))' -H 'Content-type:application/json;chareset=UTF-8' -H 'Accept:application/json' -X POST -d '{ "variantSetId": "1", "name": "", "pageToken":"" }'
{"nextPageToken":null,"callSets":[]}

```

**Comment by pgrosu** [comment]
If I specify 0 or a negative number for pageSize, I seem to get even more results:

```
curl 'http://rest.ensembl.org/ga4gh/callsets/search?fields=callSets(id,name))' -H 'Content-type:application/json;chareset=UTF-8' -H 'Accept:application/json' -X POST -d '{ "variantSetId": "1", "name": "", "pageToken":"", "pageSize": 0 }'
{"nextPageToken":10,"callSets":[{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00096","name":"HG00096","updated":1432745640000,"id":"1:HG00096"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00097","name":"HG00097","updated":1432745640000,"id":"1:HG00097"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00099","name":"HG00099","updated":1432745640000,"id":"1:HG00099"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00100","name":"HG00100","updated":1432745640000,"id":"1:HG00100"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00101","name":"HG00101","updated":1432745640000,"id":"1:HG00101"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00102","name":"HG00102","updated":1432745640000,"id":"1:HG00102"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00103","name":"HG00103","updated":1432745640000,"id":"1:HG00103"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00105","name":"HG00105","updated":1432745640000,"id":"1:HG00105"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00106","name":"HG00106","updated":1432745640000,"id":"1:HG00106"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00107","name":"HG00107","updated":1432745640000,"id":"1:HG00107"}]}
```

**Comment by pgrosu** [comment]
Funny thing, I don't even need to specify pageSize and still get the results as follows:

```
 curl 'http://rest.ensembl.org/ga4gh/callsets/search?fields=callSets(id,name))' -H 'Content-type:application/json;chareset=UTF-8' -H 'Accept:application/json' -X POST -d '{ "variantSetId": 1 }'
{"nextPageToken":10,"callSets":[{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00096","name":"HG00096","updated":1432745640000,"id":"1:HG00096"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00097","name":"HG00097","updated":1432745640000,"id":"1:HG00097"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00099","name":"HG00099","updated":1432745640000,"id":"1:HG00099"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00100","name":"HG00100","updated":1432745640000,"id":"1:HG00100"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00101","name":"HG00101","updated":1432745640000,"id":"1:HG00101"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00102","name":"HG00102","updated":1432745640000,"id":"1:HG00102"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00103","name":"HG00103","updated":1432745640000,"id":"1:HG00103"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00105","name":"HG00105","updated":1432745640000,"id":"1:HG00105"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00106","name":"HG00106","updated":1432745640000,"id":"1:HG00106"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00107","name":"HG00107","updated":1432745640000,"id":"1:HG00107"}]}
```

**Comment by pgrosu** [comment]
They seem to be using [Catalyst-Action-Rest](http://search.cpan.org/~jjnapiork/Catalyst-Action-REST-1.20/lib/Catalyst/Action/REST.pm) as their implementation, so if you prefer the data to be formatted in a nicer way, below is the commands:

```Bash
$ curl 'http://rest.ensembl.org/ga4gh/callsets/search' -H 'Content-type:chareset=UTF-8' -H 'Accept:text/xml' -X POST -d '{ "variantSetId": 1 }' | head -n20
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3142  100  3121  100    21   8214     55 --:--:-- --:--:-- --:--:--  8213
<opt>
  <data nextPageToken="10">
    <callSets name="HG00096" created="1432745640000" id="1:HG00096" sampleId="HG00096" updated="1432745640000">
      <info>
        <assembly_version>GRCh38</assembly_version>
        <variantSetName>1000 Genomes phase3</variantSetName>
      </info>
      <variantSetIds>1</variantSetIds>
    </callSets>
    <callSets name="HG00097" created="1432745640000" id="1:HG00097" sampleId="HG00097" updated="1432745640000">
      <info>
        <assembly_version>GRCh38</assembly_version>
        <variantSetName>1000 Genomes phase3</variantSetName>
      </info>
      <variantSetIds>1</variantSetIds>
    </callSets>
...
```

**Comment by mbookman** [comment]
Yeah, there seems to be a strange inter-play of parameters:

1. Body contains only `variantSetId` --> Returns results
2. Body contains `variantSetId` and empty `name` -> No results
3. Body contains `variantSetId`, empty `name`, and `pageSize` 0 -> many results

Here are examples:

#### 1. Body contains only `variantSetId` --> Returns results

```
$ curl 'http://rest.ensembl.org/ga4gh/callsets/search?fields=callSets(id,name))' -H 'Content-type:application/json;chareset=UTF-8' -H 'Accept:application/json' -X POST -d '{ "variantSetId": "1" }'
{"nextPageToken":10,"callSets":[{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00096","name":"HG00096","updated":1432745640000,"id":"1:HG00096"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},
<snip>
```

#### 2. Body contains `variantSetId` and empty `name` -> No results

```
$ curl 'http://rest.ensembl.org/ga4gh/callsets/search?fields=callSets(id,name))' -H 'Content-type:application/json;chareset=UTF-8' -H 'Accept:application/json' -X POST -d '{ "variantSetId": "1", "name": "" }'
{"nextPageToken":null,"callSets":[]}
```

#### 3. Body contains `variantSetId`, empty `name`, and `pageSize` 0 -> many results

```
$ curl 'http://rest.ensembl.org/ga4gh/callsets/search?fields=callSets(id,name))' -H 'Content-type:application/json;chareset=UTF-8' -H 'Accept:application/json' -X POST -d '{ "variantSetId": "1", "name": "", "pageSize": 0 }'
{"nextPageToken":10,"callSets":[{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},"variantSetIds":["1"],"sampleId":"HG00096","name":"HG00096","updated":1432745640000,"id":"1:HG00096"},{"created":1432745640000,"info":{"assembly_version":["GRCh38"],"variantSetName":["1000 Genomes phase3"]},
<snip>
```

**Comment by pgrosu** [comment]
So which of the fields - and parts of the data - would help you the most with the process of updating the code in this repo properly?  Then we can then just gather those initially.  As an aside, I'm not sure the filter `?fields=callSets(id,name)` is something that is implemented, which is why we seem to get the long (unfiltered) results.

**Comment by mbookman** [comment]
Thanks Paul. Right now I am able to proceed.
If I run into any blockers I'll post it here and note explicitly that I am blocked.

**Comment by pgrosu** [comment]
Cool beans! Thanks Matt :)

**Comment by mbookman** [comment]
It looks to me like there is a bug in the Ensembl `variants/search` method. Either that or the GA4GH API documentation is has omitted clearly labeling a required parameter.

Looking at the `VariantsSearchRequest` in the 0.6.0 specification:

https://github.com/ga4gh/schemas/blob/ga4gh-release-0.6.0/src/main/resources/avro/variantmethods.avdl#L64

The `variantSetId` is not listed as a required parameter.

However the implementation:

http://rest.ensembl.org/documentation/info/gavariants

lists `variantSetId` under the **Required** parameters.

This creates a problem for GABrowse here as the `variantSetId` is not stored in the client such that it can pass it in to the variants/search method. This shouldn't be too hard to update in the client, but I **thought** that the callset ID should be sufficient here.

@sarahhunt - I have two issues to file for the implementation at rest.ensembl.org. Where can I file them?

**Comment by mbookman** [comment]
Fixed by #14.
